design/decouple-ceph-version.md

```diff
@@ -0,0 +1,128 @@
+# Decoupling the Ceph version
+
+Today the version of Ceph is tied to the version of Rook. Each release of Rook releases a specific version of Ceph that is embedded in the same docker image.
+This needs to be changed such that the version of Ceph is decoupled from the release of Rook. By separating the decision of which version of Ceph will be deployed with Rook, we have a number of advantages:
+- Admins can control when they upgrade the version of Ceph. The data path upgrade will be carefully controlled by admins in production environments.
+- Admins can choose to run different versions of Ceph depending on their requirements.
+- Developers can test against any version of Ceph, whether a stable version of Luminous or Mimic, or even a private dev build.
+
+Today Rook includes Luminous, while Mimic was released several months ago. A frequently asked question from users is when we are going to update to Mimic so they can take advantage of the new features such as the improved dashboard. That question will not be heard anymore after this design change. As soon as a new release of Ceph is available, Rook users could run with it.
```

Comment:
> As soon as a new release of Ceph is available, Rook users could run with it.

The key caveat here is that it will only work if none of the behaviour Rook depends on has changed. Rook does lots of CLI-wrapping, which was safe when the Ceph version was pinned.

Given that we may rely on Rook to orchestrate Ceph upgrades (i.e. Rook knowing the magic steps for a particular major version), maybe Rook should have an explicit idea of the highest Ceph version (major release number) it knows how to run.

In practice we would hopefully get a Rook release that knows about a new Ceph release at about the same time the Ceph release comes out, so that Rook users can install the new version of Ceph via a proper upgrade process that has already been written into Rook.

Reply:
Agreed on these concerns. I'm thinking we will need a white list of ceph versions that rook supports in a given release of Rook. For example, in 0.9 rook could support versions 11 (luminous) and 12 (mimic). If nautilus were to ship before 0.9, rook could also support 13. Otherwise, 13 would fall under the the `allowUnrecognizedVersions`.  In a patch release (0.9.1), rook would plan to update the supported versions and make any related changes necessary.